File: src/fragment/trickster/spellPart.ts
import { View } from 'pixi.js';
import { Fragment } from '~/fragment/fragment';
import * as fragment from '~/fragment/fragment';
import { StructEndec, StructEndecBuilder, Optional, listOf, Endec } from 'KEndec';
import { recursive, protocolVersionAlternatives, withAlternative } from '~/endecTomfoolery';
import * as spellInstruction from '~/spellInstruction';
import { SpellUtils } from '~/SpellUtils';
import { Identifier } from '~/util';

export type SpellPartData = {
    readonly glyph: Fragment<unknown>;
    readonly subParts: ReadonlyArray<fragment.Fragment<SpellPartData>>;
};

export const COLOR = 0xaa44aa;

export const endec = (fragmentEndecs: Map<Identifier, StructEndec<Fragment<unknown>>>) =>
    recursive((self: StructEndec<Fragment<SpellPartData>>) =>
        StructEndecBuilder.of2(
            fragment
                .endec(fragmentEndecs)
                .fieldOf('glyph', (spellPart: Fragment<SpellPartData>) => spellPart.data.glyph),
            protocolVersionAlternatives(
                new Map([[1, self.listOf()]]),
                withAlternative(
                    spellInstruction.stackEndec(fragmentEndecs).xmap(
                        (instructions) =>
                            SpellUtils.decodeInstructions(instructions, [], [], Optional.empty()),
                        SpellUtils.flattenNode
                    ),
                    self
                ).listOf()
            ).fieldOf('sub_parts', (spellPart: Fragment<SpellPartData>) =>
                listOf([...spellPart.data.subParts])
            ),
            (glyph, subParts) => of(glyph, subParts.asJsReadonlyArrayView())
        )
    );


function displayOf(data: SpellPartData): () => View {
    return () => {
        throw new Error('todo');
    };
}

export function of(
    glyph: Fragment<unknown>,
    subParts: ReadonlyArray<Fragment<SpellPartData>>
): Fragment<SpellPartData> {
    const data = { glyph, subParts };
    return {
        type: new Identifier('trickster', 'spell_part'),
        display: displayOf(data),
        data: data,
    };
}

File: site/src/jsMain/kotlin/me/maplesyrum/tricksterStudio/spell/fragment/SpellPart.kt
package me.maplesyrum.tricksterStudio.spell.fragment


import me.maplesyrum.tricksterStudio.SpellUtils
import me.maplesyrum.tricksterStudio.endec.protocolVersionAlternatives
import me.maplesyrum.tricksterStudio.endec.recursiveStruct
import me.maplesyrum.tricksterStudio.endec.withAlternative
import me.maplesyrum.tricksterStudio.spell.executer.SpellInstruction
import tree.maple.kendec.StructEndec
import tree.maple.kendec.impl.StructEndecBuilder
import tree.maple.kendec.structEndecOf
import tree.maple.kendec.util.Optional

class SpellPart(var glyph: Fragment, var subParts: List<SpellPart>) : Fragment() {

    constructor(glyph: Fragment = PatternGlyph()) : this(glyph, ArrayList())

    override fun type(): FragmentType<out Fragment> {
        return FragmentType.SPELL_PART
    }

    override fun toString(): String {
        return "SpellPart"
    }

    companion object {
        val ENDEC: StructEndec<SpellPart> = recursiveStruct {
            StructEndecBuilder.of(
                Fragment.ENDEC.fieldOf("glyph") { obj: SpellPart -> obj.glyph },
                protocolVersionAlternatives(
                    mapOf(1 to it.listOf()),
                    withAlternative(
                        SpellInstruction.STACK_ENDEC.xmap(
                            { SpellUtils.decodeInstructions(it, listOf<Int>(), listOf<Fragment>(), Optional.empty<Fragment>()) },
                            (SpellUtils::flattenNode)
                        ), it
                    ).listOf()
                ).fieldOf ("sub_parts", SpellPart::subParts),
                ::SpellPart
            )
        }
    }
}
